commit ac15023c8b298c38d8def9c1ca3b21bed9fd9238
Author: Chris Boot <bootc@bootc.net>
Date:   Mon May 7 11:21:15 2012 +0100

    mach-bcm2708: fix mach/debug-macro.S so that early printk works
    
    Not sure what the original code was trying to do as it was completely
    wrong on many levels. This patch fixes the macro to return the correct
    physical and virtual addresses of the PL011 UART on the RPi.
    
    Note that you need to boot the compressed kernel (zImage) so that the
    UART is configured at boot, or your kernel will hang when it tries to
    access the UART.
    
    Signed-off-by: Chris Boot <bootc@bootc.net>

Index: linux-3.2.46/arch/arm/mach-bcm2708/include/mach/debug-macro.S
===================================================================
--- linux-3.2.46.orig/arch/arm/mach-bcm2708/include/mach/debug-macro.S	2013-07-26 19:31:51.000000000 +0000
+++ linux-3.2.46/arch/arm/mach-bcm2708/include/mach/debug-macro.S	2013-07-26 19:34:20.000000000 +0000
@@ -12,13 +12,11 @@
  *
 */
 
-		.macro	addruart, rx, tmp
-		mrc	p15, 0, \rx, c1, c0
-		tst	\rx, #1			@ MMU enabled?
-		moveq	\rx,      #0x08000000
-		movne	\rx,      #0xf8000000	@ virtual base
-		orr	\rx, \rx, #0x00200000
-		orr	\rx, \rx, #0x00001000
+#include <mach/platform.h>
+
+		.macro	addruart, rp, rv
+		ldr	\rp, =UART0_BASE
+		ldr	\rv, =IO_ADDRESS(UART0_BASE)
 		.endm
 
 #include <asm/hardware/debug-pl01x.S>
